stm32-modules/thermocycler-refresh/firmware/thermal/ads1115.cpp

```diff
@@ -0,0 +1,129 @@
+ * same physical ADC, and each can be read across multiple
+ * threads.
+ */
+struct adc_hardware {
+    // Track whether this ADC has been initialized.
+    std::atomic<bool> _initialization_started = false;
+    bool _initialization_done = false;
+    // Semaphore information for the ADC
+    SemaphoreHandle_t _semaphore = nullptr;
+    StaticSemaphore_t _semaphore_buffer = {};
+};
+
+static std::array<adc_hardware, ADC_ITR_NUM> _adc_hardware;
+
+ADC::ADC(uint8_t addr, ADC_ITR_T id) : _addr(addr), _id(id), _last_result(0) {}
+
+void ADC::initialize() {
+    bool initialization_started =
+        _adc_hardware[_id]._initialization_started.exchange(true);
+    if (initialization_started == true) {
+        // Spin until initialization is completed by other thread
+        while (initialized() != true) {
+            taskYIELD();
+        }
+    } else {
+        // Set up the semaphore for this ADC.
+        _adc_hardware[_id]._semaphore = xSemaphoreCreateMutexStatic(
+            &(_adc_hardware[_id]._semaphore_buffer));
+        configASSERT(_adc_hardware[_id]._semaphore != NULL);
+
+        // Write to the Lo and Hi threshold registers first to enable the ALERT
+        // pin
+        thermal_i2c_write_16(_addr, lo_thresh_addr, lo_thresh_default);
+        thermal_i2c_write_16(_addr, hi_thresh_addr, hi_thresh_default);
+
+        thermal_i2c_write_16(_addr, config_addr, config_default);
+
+        _adc_hardware[_id]._initialization_done = true;
+    }
+}
+
+auto ADC::read(uint16_t pin) -> ADC::ReadVal {
+    uint32_t notification_val = 0;
+    bool i2c_ret;
+    const TickType_t max_block_time = pdMS_TO_TICKS(500);
+    if (!initialized()) {
+        return ReadVal(Error::ADCInit);
+    }
+    if (!(pin < pin_count)) {
+        return ReadVal(Error::ADCPin);
+    }
+
+    if (get_lock() == false) {
+        return ReadVal(Error::ADCTimeout);
+    }
+
+    i2c_ret = thermal_arm_adc_for_read(_id);
+    if (i2c_ret == false) {
+        static_cast<void>(release_lock());
+        return ReadVal(Error::ADCTimeout);
+    }
+    // This kicks off the conversion on the selected pin
+    i2c_ret = thermal_i2c_write_16(
+        _addr, config_addr,
+        config_default | (pin << config_mux_shift) | config_start_read);
+
+    if (i2c_ret == true) {
+        // thermal_hardware.c will notify this task once the correct GPIO sends
+        // a pulse indicating ADC READY.
+        notification_val = ulTaskNotifyTake(pdTRUE, max_block_time);
+
+        if (notification_val == 1) {
+            i2c_ret =
+                thermal_i2c_read_16(_addr, conversion_addr, &_last_result);
+        }
+    }
+
+    static_cast<void>(release_lock());
+    if ((i2c_ret == false) || (notification_val == 0)) {
+        return ReadVal(Error::ADCTimeout);
+    }
+
+    return ReadVal(_last_result);
+}
+
+auto ADC::initialized() -> bool {
+    return (_adc_hardware[_id]._initialization_done == true);
+}
+
+auto ADC::get_lock() -> bool {
```

Comment:
can we use [this](https://github.com/Opentrons/ot3-firmware/blob/main/include/common/core/freertos_synchronization.hpp) from ot3 firmware, probably just by copying for now

Reply:
Oh nice, didn't know this existed!